Replace the JS object given to jl.Class by primitive IR operations.

sjrd opened the pull request:
`java.lang.Class` requires primitive capabilities that only the linker can provide for most of its core methods. This commit changes the fundamental way to provide those.

In the process, remove creation of multi-dimensional arrays from
the IR, and move it to user-space instead.

---

Previously, the contract was that the linker provided a JS object with specified properties and methods to the constructor of `jl.Class`. For example, it contained a `"name"` property in order to implement `jl.Class.getName()`. We happened to reuse the `$TypeData` instance as that object, for convenience, and to avoid more indirections.

This design grew organically over the years, and was eventually specified in Scala.js IR 1.0.0. However, it always had some shortcomings that I was not happy about.

First, the `getSuperclass()` method had to be special-cased in the Analyzer and Emitter: its very presence requires additional metadata. When it is not reachable, we do not actually provide the supposedly specified member `"getSuperclass"()` on the JS object. That assumes that only `jl.Class.getSuperclass()` actually calls that method. In turn, that assumes it never gets inlined (!), since doing so would make it unreachable, but the feature would still be required.

Second, for optimization purposes, the emitter needs direct access to the `$TypeData` stored inside `jl.Class` (for some of its Transients). There again, that assumed a field with a particular name be present in `jl.Class` in which the JS object would be stored. This implicit requirement was already leaking in the way `jl.Class` was written, in order to prevent scalac from renaming the field.

---

In this commit, we completely change the approach, and therefore the IR spec. We remove the JS object passed to `jl.Class`; its constructor now takes no arguments. Instead, we add primitives in the IR, in the form of new `UnaryOp`s and `BinaryOp`s. These new primitives operate on an argument of type `jl.Class!`, and possibly a second argument.

For example, `UnaryOp.Class_name` replaces the functionality previously offered by the `"name"` field. `BinaryOp.Class_superClass` replaces `"getSuperclass"()`.

It is up to the backend to ensure it can, somehow, implement those operations given an instance of `jl.Class`. We choose to add a magic field to `jl.Class` at the Emitter level to store the corresponding instance of `$TypeData`, and implement the operations in terms of that.

The approach has several benefits:

* It solves the two fundamental issues of the JS object, mentioned above.
* It does not require "public" (unminifiable) property names for the features; since there is no JS object spec, we can consistently minify those members.
* Several methods that were given an *intrinsic* treatment by the optimizer now fall, more naturally, under regular folding of operations.

---

Since we are changing the spec anyway, we use the opportunity to switch what is primitive about `Array.newInstance`. Previously, the overload with multiple dimensions was primitive, and the one with a single dimension delegated to it. This was a waste, since usages of the multi-dimensional overload are basically non-existent otherwise. Similarly, the `NewArray` node was designed for multiple dimensions, and the single dimension was a special case.

We now make the one-dimensional operations the only primitives. We implement the multi-dimensional overload of `newInstance` in terms of the other one. We also use that overload as a replacement for multi-dimensional `NewArray` nodes. This simplifies the treatment in the linker, and produces shorter and more efficient code at the end of the day.

---

The change of approach comes with a non-negligible cost for backward compatibility. It is entirely done using deserialization hacks. The most glaring issue is the synthesization of the multi-dimensional overload of `Array.newInstance`.

In the first big commit, we do not change the library/compiler yet, so that we can exercise the deserialization hacks.

The second commit changes the compiler and library and disable the deserialization hacks when reading IR v1.17+.

gzm0 commented:
I'm prioritizing #4988 over this one. LMK if I should change that.

gzm0 commented:
One thought I had that you can probably already answer: What prompted the decision to model this as operators rather than special purpose IR trees?

sjrd commented:
TBH, I did not like the idea to add 10 different IR nodes for unfrequent operations. IR nodes are "expensive" in terms of maintenance.

Unary and binary ops originally had the rationale that they were all together because they were *pure*, but that was broken when we made integer divisions reliably throw on `/ 0`. Additionally, `String_+` is not really pure, since we have no guarantee that the `toString()` methods are pure. `String_charAt` even introduced semantics-dependent UB handling, so the fact that we now have NPEs in ops is not truly new.

So I reasoned that what made operators still meaningful as a category is that they share the following properties:

* They have a fixed number of arguments
* They first evaluate their arguments, from left to right, irrespective of the operator
* Then they compute a result based on the value of the arguments, and not on scope or control-flow or things like that

Those properties are meaningful for manipulating the trees. For example, unnesting in `FunctionEmitter` is handled for all operators in a consistent way thanks to those properties.

Under that category, I would actually consider integrating `GetClass`, `Clone` and `IdentityHashCode` as unary ops.

Conversely, `String_+` might deserve its own IR node, because it is the only one whose result may also depend on the *state* that is accessible from the operand values (all the other ones are stateless).

gzm0 commented:
> Those properties are meaningful for manipulating the trees. For example, unnesting in FunctionEmitter is handled for all operators in a consistent way thanks to those properties.

> Under that category, I would actually consider integrating GetClass, Clone and IdentityHashCode as unary ops.

Agreed. Probably we should do that to reduce trees (and document the properties on the ops).

> Conversely, String_+ might deserve its own IR node, because it is the only one whose result may also depend on the state that is accessible from the operand values (all the other ones are stateless).

Doesn't this also apply to `Clone`? But TBH, I'm not sure that this is a meaningful criteria for exclusion. Ops can be side-effecting, and it feels that distinguishing this specific type of side-effect is kind of odd :man_shrugging:

sjrd commented:
Ah yes, that's true. It also applies to `Clone`. I had not thought about that. 

Exclusions are good if the nodes need to be treated differently in most cases. But otherwise probably not.

sjrd commented:
I was thinking about this on the side, and I find it annoying that we need to handle `null`s. In theory, since the `Class_x` operators can be used with arbitrary operands, and we don't have non-nullable `ClassType`s, the optimizer and backends must be ready to handle `null`s. In practice, however, the only occurrences of these operators uses `This()` as the (lhs) `Class` value, which is statically known to be non-null. That makes the optimizer and backends more complicated than necessary, as accidental complexity.

I contemplated adding non-nullable `ClassType`s in the IR, but that quickly looked daunting at this time.

However, what if the `ClassDefChecker` only allowed the (lhs) operand to `Class_x` operators to lexically be a `This()` node? After the optimizer, that would be relaxed (conditioned on the `allowTransients` flag, probably). This way, even without non-nullable types in the IR, we have a static guarantee that the operand to `Class_x` is non-null, and that means we don't need to specify and implement what happens when it's `null`.

If we do introduce non-nullable types in the IR in the future, we could then naturally remove that restriction and only demand a non-null `jl.Class` for the operand.

@gzm0 WDYT about this idea? Does it seem acceptable?

sjrd commented:
I have implemented the above alternative as an additional commit (second-to-last). If we go for it, it should be squashed with the main commit, since it undoes a number of changes of the main commit.

sjrd commented:
Rebased.

sjrd commented:
Rebased.

gzm0 commented:
It is a pity that `jl.Class` has actual implementations. Otherwise it feels like it might make sense to model values of `jl.Class` as an implementation defined primitive type.

- This would resolve the nullability issue.
- `jl.Class` would become a hijacked class.
- We wouldn't have to do the field / constructor patching.

My gut feeling is that restricting the operands lexically goes a bit far, especially if we cannot maintain the property post optimizer. This will force us to either:
- Never run the IR checker post optimizer, or
- Implement non-nullable types in the IR.

I don't think the first one is something we want to commit to. So if we are to commit to the second one, we might just do it? (It would not surprise me if this would lead to various simplifications in the backends, especially w.r.t. `AssumeNotNull`).

Looking `ir.Types`: things we'd need for non-nullable IR types:
- Add a `nullable: Boolean` field to `ClassType` and `ArrayType`
- Adjust `Types#isSubtype`
- Disallow zeros of non-nullable class types / array types
- Per consequence disallow fields of non-nullable class types / array types
- A full review of `ir.Trees` for the types they generate (other stuff can be improved as we go IMO).

sjrd commented:
> My gut feeling is that restricting the operands lexically goes a bit far, especially if we cannot maintain the property post optimizer. This will force us to either:
> 
>     * Never run the IR checker post optimizer, or
> 
>     * Implement non-nullable types in the IR.

Not necessarily. We can also have the optimizer systematically insert an `AssumeNotNull` there, and the IR checker would allow both a `This` or an `AssumeNotNull`. If run after the optimizer, the optimizer would have to know about `Transient`s anyway.

sjrd commented:
Rebased, including adding changes in the Wasm backend (the rest is unchanged).

gzm0 commented:
Have we considered adding an (opaque) type data primitive type to the IR?

The type would be non-nullable and the argument type to the class operations. IIUC this would also allow `jl.Class` to become more normal again. (IIUC the only issue we'd have to solve is the zero of the `data` field, but maybe we can hack to allow null there).

sjrd commented:
Ah, that is an interesting take. The typed closures PR shows that it is indeed possible to have IR types that don't have any interop semantics, by not being subtypes of `any`. I'll try that.

sjrd commented:
Meh, that's not so great. In addition to introducing more difficulties to handle `null`s, it reintroduces the problem that the linker must know about the `data` field of `jl.Class`. For example to constant-fold `classOf[C].getName()`, which becomes `name(classOf[C].data)`, we must be able to make the connection that `.data` is necessarily the `TypeData` value corresponding to `C`. So, not an improvement IMO.

sjrd commented:
FYI, I'm getting somewhere with nullable types in the IR, but I'm not done yet (the optimizer is not type-preserving yet, so Wasm+optimizer generates invalid Wasm code):
https://github.com/sjrd/scala-js/compare/optimizer-with-webassembly...sjrd:scala-js:nullable-types-in-ir

sjrd commented:
True nullable types in the IR available at #5018.

sjrd commented:
I don't think that would work. Hijacked classes need a corresponding primitive type. Otherwise there's no way to implement their methods. For example, most methods of `jl.Integer` work on `this.asInstanceOf[Int]`. Even `jl.String` basically works off of `this.asInstanceOf[string]`, although that's hidden in the hard-coded IR for `charAt` and `length`.

If we want to make `jl.Class` a hijacked class, we need to add a corresponding primitive type. Let's say it's `TypeDataType` (`ClassType` is taken ^^). Now we don't have a user-level Scala type that compiles to that primitive IR type, so we need compiler magic to fill in the bodies of `getName()` and similar anyway. In the end I don't think we gain anything, but we add a lot of complexity that wasn't there before (new primitive type to handle everywhere).

The boxing is non-optimal, but we've lived with it so far. AFAICT, we *could* equally get rid of the boxing without making `jl.Class` a hijacked class. It's a question of how we encode things in the backends.

sjrd commented:
Rebased to leverage the non-nullable reference types. Now the (first) argument of `Class_x` operations must be a `ClassType(jl.Class, nullable = false)`.

tanishiking commented on javalib/src/main/scala/java/lang/reflect/Array.scala:
```diff
@@ -17,18 +17,34 @@ import scala.scalajs.js
+
+    var componentType2 = componentType
+    var i = 1
+    while (i != dimensions.length) {
+      componentType2 = newInstance(componentType2, 0).getClass()
```
[note]
For example, `componentType = Int` and `dimensions = [2,2,2]`.
`componentType2` will be `Array[Array[Int]]`, that is the component type of outer most Array.

tanishiking reviewed (commented):
Review except for `Serializers.scala` and `OptimizerCore.scala`. It's great to see we could remove JS object for `Class#data`, and `closure` usages in `Class#data` Wasm backend :)

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala:
```diff
@@ -1407,6 +1449,9 @@ private class FunctionEmitter private (
       case Double_>  => wa.F64Gt
       case Double_>= => wa.F64Ge
+
+      case Class_isAssignableFrom => wa.Call(genFunctionID.isAssignableFromExternal)
```
Ah indeed, we can do that now that the `lhs` is guaranteed to be non-null. Previously that would have trapped too early, without evaluating the `rhs`. Good point.

gzm0 commented:
> The boxing is non-optimal, but we've lived with it so far. AFAICT, we could equally get rid of the boxing without making jl.Class a hijacked class. It's a question of how we encode things in the backends.

My concern isn't the boxing but the magic field. I don't like it :P

> If we want to make jl.Class a hijacked class, we need to add a corresponding primitive type. Let's say it's TypeDataType (ClassType is taken ^^). Now we don't have a user-level Scala type that compiles to that primitive IR type, so we need compiler magic to fill in the bodies of getName() and similar anyway.

Do we actually need a new primitive type or just new primitive operations? The latter we are planning to introduce with this PR anyways. 

But I agree that the boxing in and of itself is not a concern. So I'll proceed to reviewing this in full. I suppose the magic field isn't that bad (I have not given this a full look yet).

sjrd commented:
> > The boxing is non-optimal, but we've lived with it so far. AFAICT, we could equally get rid of the boxing without making jl.Class a hijacked class. It's a question of how we encode things in the backends.
> 
> My concern isn't the boxing but the magic field. I don't like it :P

Does it help alleviate your concerns if you think of it as conceptually a global `Map[Class, Data]`, that we happen to "optimize" as a magic field?

With `identityHashCode`s we choose the map implementation (mostly because otherwise we need a field in *every* class, usually for nothing). For the class data we choose the field implementation, because we know we'll need the data for every `Class` anyway, so there is no downside.

> > If we want to make jl.Class a hijacked class, we need to add a corresponding primitive type. Let's say it's TypeDataType (ClassType is taken ^^). Now we don't have a user-level Scala type that compiles to that primitive IR type, so we need compiler magic to fill in the bodies of getName() and similar anyway.
> 
> Do we actually need a new primitive type or just new primitive operations? The latter we are planning to introduce with this PR anyways.

If we only add primitive operations and no primitive type, then by definition it's exactly what we have in this PR, and not a hijacked class. Hijacked classes are, by definition, classes that have no instances of theirs, and instead their instances are primitive values.

Primitive operations that are class-specific are not exclusive to hijacked classes. For example, `Clone` is specific to `jl.Cloneable`, and `UnwrapFromThrowable` is specific to `jl.Throwable`.

gzm0 commented:
> Does it help alleviate your concerns if you think of it as conceptually a global Map[Class, Data], that we happen to "optimize" as a magic field?

A bit actually, surprisingly :laughing: 

> If we only add primitive operations and no primitive type, then by definition it's exactly what we have in this PR, and not a hijacked class. Hijacked classes are, by definition, classes that have no instances of theirs, and instead their instances are primitive values.

I think I understand where I wasn't clear: I understand that we need a new primitive value (the typedata in this case). However, do we really need an explicit type in the IR for this value? All the operations that generate it (`GetClass`, `ClassOf`) can immediately type their result value as the respective hijacked boxed type.

The behavior of hijacked classes that I find interesting here is that public instance methods are emitted statically and dispatch taken care of by runtime code.

(just TBC: either way I believe we should go forward with this PR, but I think it makes sense to think this through, even just to clear up / align understanding of hijacked classes).

sjrd commented:
> I think I understand where I wasn't clear: I understand that we need a new primitive value (the typedata in this case). However, do we really need an explicit type in the IR for this value? All the operations that generate it (`GetClass`, `ClassOf`) can immediately type their result value as the respective hijacked boxed type.
> 
> The behavior of hijacked classes that I find interesting here is that public instance methods are emitted statically and dispatch taken care of by runtime code.

Ah, I see. That's not a defining feature of a hijacked class, though. It's an emitter implementation detail. For example, not all hijacked classes have dedicated dispatch (`Long`s on both backends, `Char`s on Wasm). We currently do not have any example of a non-hijacked class that uses dedicated dispatch, but it could happen. We do have `Throwable` which is a bit magic in the JS emitter, since we rewire its superclass to be `js.Error` instead of `jl.Object`. Yet `Throwable` is not a hijacked class.

Hijacked classes are an IR-level concept, which stays true no matter how you choose to encode them in a backend. You even have to deal with that concept if you write an IR interpreter. And their core defining feature is, as I mentioned, that their is an associated IR primitive type that embodies their instances. This has implications on the IR type system, for example, which is pretty deep as far the IR spec is concerned.

Now all that aside, your question remains: should we *encode* `jl.Class`es as their type data in the backend and implement them using dedicated dispatch, *despite* them not being hijacked classes? I see the following trade-offs:

| Dispatch         | Boxing                                                   |
|------------------|----------------------------------------------------------|
| No dereferencing | Dereference to access concrete data                      |
| Cost of dispatch | No cost of dispatch (especially for all the other types) |
| No boxing        | Boxing (but there is O(1) box per type data)             |

Dedicated dispatch is a cost, since it does not follow the straight-line virtual dispatch codegen. This is worse in JS because it affects all the *other* types: we test our special cases first, then fallback on regular method call (because testing for the regular case requires *2* `instanceof` tests -- remember `Throwable`). On Wasm we first test our `jl.Object`s, then the special cases (because the special cases test are super expensive -- they do JS interop).

Given that there is O(1) box per type data, and that `Class` operations are not *that* frequent, IMO the better trade-off remains to use boxing.

sjrd commented:
Rebased on top of the checked semantics.

gzm0 commented:
> Now all that aside, your question remains: should we encode jl.Classes as their type data in the backend and implement them using dedicated dispatch, despite them not being hijacked classes?

Right. Specific decision in the BE aside, that's a good thing: it validates the FE design of this PR: A BE can make either decision. 

I'll have a look at dispatch specifically; IIUC this concerns dispatch of toString, equality, hashCode and instance tests, correct?

gzm0 commented:
Regarding Array instantiation:
- Why do we keep the indirection via jl.Class? Is this a remainder of the time where LHS had to this?
- Are we sure removing support for multi-dimensional arrays is worth it? I agree that would we do it again we'd do it that way. But it feels how much we need to hack in the Serializer is considerable and with a Scala Array as RHS of the op, it feels that backends can still optimize statically if the loop is a performance issue (obviously we need to ditch J's arrays in the interface, but that's much easier IIUC).

sjrd commented:
> I'll have a look at dispatch specifically; IIUC this concerns dispatch of toString, equality, hashCode and instance tests, correct?

There's also the `GetClass` node and the related transient `ObjectClassName` (I think that's what it's called).

> Regarding Array instantiation:
> 
>     * Why do we keep the indirection via jl.Class? Is this a remainder of the time where LHS had to this?

It's the easiest way to get a non-nullable `jl.Class`: by forcing it into a `this`.

The alternative is to also add a new IR feature to check null of an expression in a way that actually gives us a non-nullable result type. We have `GetClass(x)` but that only works in statement position. Two designs that I've thought about are: 
* A unary op `Any_checkNotNull` (with a result type that depends on its argument type, which would be new for `UnaryOp`).
* Allow non-nullable target types for `AsInstanceOf`, which would thank throw an NPE if the operand is `null`.

>     * Are we sure removing support for multi-dimensional arrays is worth it? I agree that would we do it again we'd do it that way. But it feels how much we need to hack in the Serializer is considerable and with a Scala Array as RHS of the op, it feels that backends can still optimize statically if the loop is a performance issue (obviously we need to ditch J's arrays in the interface, but that's much easier IIUC).

I'm not completely sure. It's a lot of complexity in one place versus a bit of complexity everywhere. I can try the alternative in another branch to compare.

gzm0 commented:
> A unary op Any_checkNotNull (with a result type that depends on its argument type, which would be new for UnaryOp).

:) I was thinking that now that we have nullable types in the IR, it feels like the `CheckNull` transient shouldn't be a transient. 

I actually was wondering what would happen if we changed the specification of (select) IR nodes to not take nullable values anymore and push the burden of inserting null-checks on the compiler.

sjrd commented:
> I actually was wondering what would happen if we changed the specification of (select) IR nodes to not take nullable values anymore and push the burden of inserting null-checks on the compiler.

I think what will happen is mostly a lot more nodes in our trees, resulting in more memory pressure and slower transformations. 😉

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -1028,6 +969,31 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
+   *
+   *  Like `getClassOf`, but returns `null` if the argument is `null`.
+   */
+  private def genGetClassOfNullable()(implicit ctx: WasmContext): Unit = {
```
This seems to be only used by `Class_componentType`. Wouldn't it be better (code-size wise) to move the get's of the class data also into this helper?

IMO this would also make it clearer what this is used for / why this is here.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -2100,141 +2017,123 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
 
-  /** `getComponentType: (ref typeData) -> (ref null jlClass)`.
-   *
-   *  This is the underlying func for the `getComponentType()` closure inside class data objects.
```
Tying to my comment above: I think we should keep this.

sjrd replied:
Huh, yes, indeed. That's much better.

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/CoreWasmLib.scala:
```diff
@@ -2100,141 +2017,123 @@ final class CoreWasmLib(coreSpec: CoreSpec) {
+      // case _ => array.new_default anyrefArray; struct.new ObjectArray
+      val arrayTypeRef = ArrayTypeRef(ClassRef(ObjectClass), 1)
+      fb += ArrayNewDefault(genTypeID.underlyingOf(arrayTypeRef))
+      fb += StructNew(genTypeID.forArrayClass(arrayTypeRef))
```
Regarding complexity of `Class_newArray`: I feel this is actually more complex now, because `NewArray` and `Class_newArray` are not aligned in what they do.

If `lenghts` was an array here, we could simply get the type data and call `newArrayObject`. We probably wouldn't even need a core helper for it.

sjrd replied:
Sure, but part of the appeal of the switch is to make the `newInstance(length)` faster by not having to allocate an array, because it's the common case (like ... the other one is never used in practice 🙄).

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
I'll hold off reviewing this, until we have settled the new array thingy.

gzm0 commented:
Regarding array sizes: I feel as long as we have `NewArray` that supports multi dimensional arrays, the complexity the reflective version adds will always be very limited. IMO its highly unlikely that we'll unroll multi dimensional `NewArray`s statically; so as long as they are around, the reflective case can always be implemented in terms of that one in the backends (retrieving the component type reflectively needs to happen either way, so that's independent).

sjrd commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -2188,17 +2188,36 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
             /* Hijack the bodies of String.length and String.charAt and replace
              * them with String_length and String_charAt operations, respectively.
              */
             methodName.name match {
```
Maybe it's a bit odd. I guess we do it this way because we have the `Symbol`s of the classes anyway, and comparing them is a reference equality test. For the methods, we would have to look up all their symbols only to successfully compare them once, so it would be a lot of wasted effort.

sjrd commented on ir/shared/src/main/scala/org/scalajs/ir/Trees.scala:
```diff
@@ -446,6 +505,8 @@ object Trees {
       case String_charAt =>
         CharType
+      case Class_cast | Class_newArray =>
+        AnyType
```
That cannot happen, indeed. I think it's a leftover of before we rebased on top of the non-nullable reference types.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala:
```diff
@@ -326,8 +327,20 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
+          val dataParam = js.ParamDef(fileLevelVarIdent(VarField.data))
+          val createDataField = js.Assign(
+              js.DotSelect(js.This(), genArrayClassPropertyForDef(ArrayClassProperty.data)),
+              dataParam.ref)
```
Renamed to `SyntheticProperty` in a separate (first) commit.

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:
```diff
@@ -3153,41 +3137,8 @@ private[optimizer] abstract class OptimizerCore(
-
-            contTree(finishTransformBindings(
-                bindingsAndStats, StringLiteral(nameString)))
-
```
Yes, but also `jl.Class` would always be considered to be instantiated, which would cancel some of the benefits of having `ObjectClassName` at all. Currently this optimization allows `jl.Class` to be entirely dead-code-eliminated if neither `GetClass` nor `ClassOf` are used.

(but perhaps if it was a hijacked class, it would not produce so much more code anyway; not sure)

sjrd replied:
No actually that's not even true, because `GetClass` returns a nullable `jl.Class`. So it would still be reified at least for a `CheckNotNull` before continuing to `Class_name`.

sjrd commented:
So re-working on the `newArray` implementation made me realize that we've always been ignoring the `IllegalArgumentException`s that are supposed to be thrown in the following conditions:
* For both overloads: if the `cls` is in fact `classOf[Unit]` (ironically there was https://github.com/scala-js/scala-js/pull/5018/commits/a79ce3f945968b77d2d7815e3be096e4a8fc3ceb which prevented the optimizer from messing it up, but the run-time code behind is still broken)
* For the multi-dimensions only: when the array is *empty* (rather than `null`)

Given that it has always been broken, IMO it's best to fix those things after we land either #5031 or this PR. That said, it adds to the list of things both variants of the operation will need to deal with. On the one hand, it gives one more advantage to the single-dimension variant because there's only one failure mode to deal at the operation level; on the other hand, since both failure modes throw the same exception, the checks could maybe be reused better in the multi-dimensions variant. 🤷‍♂️

gzm0 commented:
> the checks could maybe be reused better in the multi-dimensions variant. 🤷‍♂️

In case we go with single dimension in IR only, I think we can live with the class check being over-performed if that makes the implementation of multi dimension new array easier (which it likely does).

gzm0 commented on linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/SyntheticProperty.scala:
```diff
@@ -14,33 +14,39 @@ package org.scalajs.linker.backend.emitter
-/** Represents a property of one of the special `ArrayClass`es.
+/** Represents a synthetic property of some Scala classes.
+ *
+ *  A synthetic property not exist as `FieldDef` or `MethodDef` in the IR, but
```
Typo: ... property **does** not exist

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Trees.scala:
```diff
@@ -339,10 +356,42 @@ object Trees {
+   *  throw an `ArithmeticException` when their right-hand-side is 0. That
+   *  exception is not subject to undefined behavior.
+   *
+   *  `String_charAt` throws string index out bounds.
```
`StringIndexOutOfBoundsException`?

gzm0 reviewed (changes requested):
Nits discovered on a re-review.

RE dispatch: I think we should go with the magic `data` field for now to get this moving. We can still decide to change the dispatch if we believe it is an advantage.

RE single / multi dimensions: See my other comment.

sjrd commented:
> What happens if we introduce a `newArray(tpe: Class[_], dimensions: Array[Int])` helper function for the deserializer in the linker private library and use it both for the deserialization hack proposed here and for `NewArray` with more than one dimension?

Huh. I don't think we have precedent for the deserialization hacks to rely on a method in the linker library. The hacks turn previously valid IR into newly valid IR, regardless of the linker. The linker library is an implementation detail of the linker. It is even possible for an alternative linker not to have a linker library at all.

I think we could conceivably rewrite multi-dimension `NewArray` into a call to `j.l.r.Array.newInstance`, because that one is at least publicly specified.

sjrd commented:
Added two commits to be squashed that also rewrite multi-`NewArray` into calls to `jlr.Array.newInstance`.

I kept the single-dim `NewArray` after all, because its `tpe` is better than the result of `Class_newArray`. A number of optimizations would regress if we rewrote to `Class_newArray`.

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1493,6 +1501,243 @@ object Serializers {
+          MethodDef(method.flags, newName, method.originalName,
+              Nil, NoType, Some(newBody))(
+              method.optimizerHints, method.version)
+        } else if (methodSimpleNameString == "newArrayOfThisClass") {
+          val newName = MethodIdent(MethodName(methodName.simpleName,
+              List(IntRef), ClassRef(ObjectClass)))(method.name.pos)
+          val lengthParam = ParamDef(LocalIdent(LocalName("length")),
+              NoOriginalName, IntType, mutable = false)
+          val newBody = BinaryOp(BinaryOp.Class_newArray, thisJLClass, lengthParam.ref)
+          MethodDef(method.flags, newName, method.originalName,
+              List(lengthParam), AnyType, Some(newBody))(
+              method.optimizerHints.withInline(true), method.version)
```
Would you be very opposed to expose `CheckNull` as a `UnaryOp`? Then this (and the strange private method in `Class`) would go away.

sjrd replied:
No I think it definitely makes sense. There's a bit of open design questions for that, though: 

* If we make it a `UnaryOp`, we cannot determine the `tpe` of a `UnaryOp`'s `op` alone anymore. It now depends on the type of the argument. Should this be grounds to make it its own node instead? Not sure. (Note that `Clone` will have the same issue if we want to turn it into a `UnaryOp`.)
* Instead of a new kind of operation, should it be merged into the behavior of `AsInstanceOf`? If the target type is not nullable (but also not a primitive because that's an unboxing operation) and the argument's type is nullable, then the behavior would include a null check. That has the useful side effect that we can express an `AsInstanceOf` from non-nullable to non-nullable as well, basically for free. However it's a more involved change. 

Do you have any opinion about these alternatives?

gzm0 replied:
> Should this be grounds to make it its own node instead?

I'm not sure. It feels like that change would still be extremely local. All usage sites are going to look at `tpe` anyways if they need the type.

> Instead of a new kind of operation, should it be merged into the behavior of AsInstanceOf?

I think conceptually this would be cleaner, yes. I have been fiddling with this and a couple of challenges surface already:

- We need more optimization in the emitter stage: Checking for null and for a type are very different at the emitter level. So always doing both is wasteful.
- If we naively replace `CheckNotNull(expr)` with `AsInstanceOf(expr, expr.tpe.toNonNullable)` we get undefined references (at the GCC level) for `$as_jl_Long` and `$as_jl_Throwable`. The former is probably the fault of the optimizer (runtime long substitution is the last thing IIUC that prevents us from running IR checking post optimizer). The latter is probably the fault of the emitter (JS throwable replacement).
- Side-effect checking for `AsInstanceOf` under semantics becomes more complicated: We now depend on two semantics (asInstanceOf and NPE). Again, if we do the checking naively (require both semantics to be unchecked), we'll almost certainly regress in code size.

So all-in-all, I'm not sure we're better of with the `AsInstanceOf` option (but maybe more testing is required). I  think what we should definitely do is allow non-nullable types in `AsInstanceOf` if the target expr is already non-nullable (but IIUC that is a trivial change).

sjrd replied:
I agree with the analysis. It seems merging with `AsInstanceOf` causes more issues than it solves.

I made the unary op version in #5037.

gzm0 reviewed (changes requested):
I think we should proceed with the variant with the simplified NewArray: IMO the deser hacks do not get much worse, but we get more simplification downstream.

I haven't finished in-detail review yet, but have one comment RE CheckNull.

tanishiking commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1493,6 +1531,238 @@ object Serializers {
+
+          MethodDef(method.flags, method.name, method.originalName,
+              method.args, method.resultType, Some(newBody))(
+              method.optimizerHints.withInline(true), method.version)
```
It seems marking `getSimpleName` and `computeCachedSimpleNameBestEffort` inline flagged too (am I correct?), is that okay? I don't see those methods`@inline` in Class.scala.

sjrd replied:
Good catch! Fixed.

sjrd commented:
Rebased on top of #5037, being optimistic that we're going to opt one way or another for a version based on a check-not-null operation.

sjrd commented:
Squashed :)

sjrd commented on linker/shared/src/main/scala/org/scalajs/linker/checker/IRChecker.scala:
```diff
@@ -471,17 +474,25 @@ private final class IRChecker(unit: LinkingUnit, reporter: ErrorReporter) {
+            IntType
+          case Class_isInstance | Class_cast =>
+            AnyType
+          case Class_isAssignableFrom =>
+            ClassType(ClassClass, nullable = true)
```
Now that we have the ability to insert a `CheckNotNull`, should we also require a non-nullable `jl.Class` as the *rhs* of `isAssignableFrom`?

It's not as critical as the lhs, because the NPE naturally comes after evaluating both arguments, which is not true for the lhs. But it would probably be more consistent.

WDYT?

sjrd replied:
This is quite a bit better with that change, IMO, so I did it.

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1493,6 +1530,253 @@ object Serializers {
+                      outermostComponentType.ref,
+                      GetClass(Apply(EAF, This()(ClassType(ReflectArrayModClass, nullable = false)),
+                          MethodIdent(newInstanceSingleName),
+                          List(outermostComponentType.ref, IntLiteral(0)))(AnyType))
```
Is it deliberate that you don't use `BinaryOp.Class_newArray` here?

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1493,6 +1530,253 @@ object Serializers {
+
+        val length = varDef("length", IntType, ArraySelect(dimensions.ref, offset.ref)(IntType))
+        val result = varDef("result", AnyType,
+            Apply(EAF, ths, MethodIdent(newInstanceSingleName), List(componentType.ref, length.ref))(AnyType))
```
Is it deliberate that you don't use `Class_newArray` here?

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1493,6 +1530,253 @@ object Serializers {
+        val result2 = varDef("result2", ArrayType(objectArrayTypeRef, nullable = true),
+            AsInstanceOf(result.ref, ArrayType(objectArrayTypeRef, nullable = true)))
+        val innerComponentType = varDef("innerComponentType", jlClassType,
+            Apply(EAF, componentType.ref, MethodIdent(getComponentTypeName), Nil)(jlClassType))
```
Also here: Is it deliberate that you don't use the primitive unary op?

gzm0 reviewed (approved):
Just some questions regarding the deser hack.
